Plane: implement manual forward throttle for QACRO, QSTABILIZE and QHOVER

kd0aij opened the pull request:
1 new parameter: Q_FWD_THR_MAX and a new RCx_OPTION: FWD_THR
If FWD_THR is set for an RC channel, enables manual forward throttle control via that channel
Manual throttle ranges from zero to Q_FWD_THR_MAX with 1 being full throttle.

also adds rudder control from stick input for QACRO mode

replaces #11548

Question: should there be a flight mode in which forward throttle is somehow mixed in from throttle or pitch stick instead of using a separate channel? The stock Convergence has a mode in which the pitch stick controls only forward tilt, but this means altitude is controlled only by throttle, with pitch held near zero. The result is sluggish control of altitude.
Suggestion from @Hwurzburg : mix in forward tilt from throttle when throttle stick is above hover thrust

kd0aij commented:
Flight test with Convergence tiltrotor, pitch up to hold zero airspeed while increasing forward throttle. 
At low airspeed, there is a limit for forward tilt beyond which roll control is degraded (note that this is not an issue for non-tiltrotors):

![fwd_thr_limit](https://user-images.githubusercontent.com/2300221/61072945-93857480-a3d1-11e9-9aea-35db15d6c4b2.png)
 
This was with Q_FWD_THR_MAX = 0.3, and control stability may be dependent on other parameters also, such as Q_TILT_YAW_ANGLE and Q_TILT_MAX.

kd0aij commented:
Flight tested on E-Flite Convergence and mini-Convergence.

forward throttle and QACRO rudder control tested in SITL with RF8 BigStik Quadplane

kd0aij commented:
rebased on master

kd0aij commented:
takeoff and landing in QHOVER mode with 20 mph wind, using forward throttle to compensate for wind.  
Wind was mostly from the south and heading was into the wind.
Flight controller was Mateksys F405-Wing
![T L_in_wind](https://user-images.githubusercontent.com/2300221/65353172-b12d4200-dba9-11e9-8aaf-372ee478aa78.png)

kd0aij commented:
force-pushed a whitespace fix

kd0aij commented:
@IamPete1 I think you reviewed an earlier version of this PR.  This one is cleaned up a bit, and I just added a new commit to allow testing the forward throttle and vectored yaw tilt functions while disarmed. The fact that they didn't work while disarmed caused a new user quite a bit of frustration.

kd0aij commented:
SITL demo: https://discuss.ardupilot.org/t/e-flite-mini-convergence-with-ardupilot/46015/4

IamPete1 commented on ArduPlane/quadplane.cpp:
```diff
@@ -461,6 +460,20 @@ const AP_Param::GroupInfo QuadPlane::var_info2[] = {
+    // @Description: RC channel for manual control of VTOL forward throttle.
+    // @Range 0 16
+    // @RebootRequired: True
+    AP_GROUPINFO("FWD_THR_CH", 16, QuadPlane, fwd_thr_chan, 0),
```
I think its better to try and move to a RC_Option for this, you define the option and just use it as a label rather than do anything with it in RC_Channels, see how the Mainsail input (207) works on rover

kd0aij replied:
Thanks for the suggestion; eliminates that parameter and saves a line of code, I think.

IamPete1 commented on ArduPlane/quadplane.cpp:
```diff
@@ -443,7 +443,6 @@ const AP_Param::GroupInfo QuadPlane::var_info2[] = {
     // @Increment: 1
     // @User: Standard
     AP_GROUPINFO("ACRO_YAW_RATE", 13, QuadPlane, acro_yaw_rate, 90),
-
```
accidentally removed line

kd0aij replied:
several accidental whitespace changes. I can't get meld to show these for some reason.
Will have to start using "git add -p" religiously :)

IamPete1 commented on ArduPlane/tiltrotor.cpp:
```diff
@@ -90,29 +90,48 @@ void QuadPlane::tiltrotor_continuous_update(void)
+    if (!assisted_flight &&
+        (plane.control_mode == &plane.mode_qacro ||
+         plane.control_mode == &plane.mode_qstabilize ||
+         plane.control_mode == &plane.mode_qhover)) {
```
This stuff is already checked in  forward_throttle_pct() I think, so you can probably just call that without checking first. The assisted_flight check is the only one not checked there think

kd0aij replied:
Since this returns early, I think that would change behavior.

IamPete1 replied:
ah, I see what you mean, my mistake

tridge commented on ArduPlane/quadplane.cpp:
```diff
@@ -2747,22 +2757,42 @@ void QuadPlane::Log_Write_QControl_Tuning()
 {
+    // disable forward motor if disarmed, spooled down or not in a VTOL mode
+    if (!in_vtol_mode() ||
+        (SRV_Channels::get_output_channel_mask(SRV_Channel::k_throttle) &&
```
i don't think we should check for k_throttle here

kd0aij replied:
added a boolean arg to forward_throttle_pct, as suggested

kd0aij commented:
rebased and added new parameter Q_FWD_THR_MIX which adds in forward throttle when above hover thrust
added throttle is in addition to the demand from rc_fwd_thr_ch

kd0aij commented:
replaced by #14520
